### PR TITLE
Add pgx driver package

### PR DESCRIPTION
**Pull request by dzbee**
Closes https://github.com/golang-migrate/migrate/issues/512

Changes:

- Replace `lib/pq` dependency with `pgx/stdlib` in the postgres driver
- Replace use of `pq.QuoteIdentifier` with private `quoteIdentifier` function in package
- Slight change around parsing of cursor position from `PgError` (which is an `int32` in `pgx.PgError` rather than `string` as in `pq.PgError`)

This PR would slightly change error reporting to include the SQL error code in the error message. If this is considered a backwards compatibility breaking change, I could add some parsing to omit the code so that this would only be a minor update rather than a major update, but I can't imagine that including additional information in the error message merits a major bump in the first place.

### CODE REVIEW

File: database/postgres/postgres.go
Comment (dhui) [inline database/postgres/postgres.go:436]: Use this implementation and reference it: https://github.com/lib/pq/blob/v1.9.0/conn.go#L1611

**Review by dhui — COMMENTED** [review]
Thanks for the PR! Good to know that the smoke tests pass!
The different error messages may break existing postgres users. e.g. if they were handling said errors
Also, I believe `WithInstance()` may behave differently using `pgx/stdlib` instead of `lib/pq`, forcing users to migrate to `pgx`.

I'll hold off on merging this until there is more demand for it.

**Comment by dzbee** [comment]
@dhui With regards to concerns about creating conflicts for `lib/pq` users in the near-term, I wonder if you would consider relaxing this intention while the future of `lib/pq` vs `pgx` (or another replacement) remains unresolved:

>...there should really just be one driver for a database...

To ease the transition from `lib/pq` to `pgx` I could split these changes out into a `pgx.go` and register the driver with that name instead of `"postgres"`. Indeed `lib/pq` in some sense "owns" the `"postgres"` driver identifier whereas `pgx/stdlib` uses `"pgx"` to avoid conflicts on initialization in cases where both modules are imported.

Let me know what you think about this alternative.

**Comment by dhui** [comment]
Registering `pgx` as a separate `migrate` db driver is a great idea! e.g. following `pgx/stdlib`'s lead
That way users can pick which postgres db driver to use. I'd hold off on recommending using `pgx` until the more of the community has moved over.

File: database/pgx/README.md
Comment (dhui) [inline database/pgx/README.md:1]: Rename to pgx

File: database/pgx/README.md
Comment (dhui) [inline database/pgx/README.md:3]: Update the DSN

File: database/pgx/README.md
Comment (dhui) [inline database/pgx/README.md:35]: Is this also true for pgx?

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:53]: Thoughts about renaming this to `Pgx` or keeping `Postgres`? Wondering if `pgx.Pgx` stutters or if `pgx.Postgres` is confusing.

**Review by dhui — CHANGES_REQUESTED** [review]
Tests are failing

File: database/pgx/pgx.go
Comment (dzbee) [reply database/pgx/pgx.go:53]: I left it as is simply because it is ultimately a Postgres driver, but I defer to your judgement if it's confusing or not sufficiently distinguished from the existing Postgres driver.

File: database/pgx/README.md
Comment (dzbee) [reply database/pgx/README.md:35]: Yes it is!

File: database/pgx/pgx.go
Comment (dhui) [reply database/pgx/pgx.go:53]: Let's keep `pgx.Postgres`

File: go.mod
Comment (dhui) [inline go.mod:30]: Looks like [v4.10.1](https://github.com/jackc/pgx/releases/tag/v4.10.1) is the latest release. Any reason why we don't use that?

Comment (dzbee) [reply go.mod]: That was simply a mistake resulting from the fact that `pgx` didn't support Go modules until v4. Therefore I was pulling the latest release prior to the introduction of modules into the project. The major version includes some breaking changes for the code I'd written before so I've included some updates with a version bump in https://github.com/golang-migrate/migrate/pull/517/commits/def44bae275b91be5eafe5955ebe292c459f5080.

Most significantly

Comment (dzbee) [reply go.mod]: Most significantly, in v3 the connection string parsing didn't include any checks on the content scheme name (see https://github.com/jackc/pgx/blob/a413de981978e4100d5736af7aaa2b392511612f/conn.go#L1246 and https://github.com/jackc/pgx/blob/a413de981978e4100d5736af7aaa2b392511612f/conn.go#L979) whereas now the `postgres` or `postresql` schemes are explicitly required in the connection string (see https://github.com/jackc/pgconn/blob/5daa019e4eb52df3409ebf17c83116b7c0e827e5/config.go#L199). Therefore I've introduced a string replacement so that after the `pgx` scheme is used to identify the appropriate driver in migrate, it's replaced with the appropriate `postgres` scheme for actually opening the connection.

File: database/pgx/pgx.go
Comment (dhui) [inline database/pgx/pgx.go:127]: Avoid manipulating URLs via string operations. Instead, set `purl.Scheme`.
I like the documentation in the comments though. Let's keep that!

**Review by dhui — CHANGES_REQUESTED** [review]
What are the breaking changes in connection and error handling. It just looks like the connection errors were moved to another package.

**Comment by dzbee** [comment]
Connection errors were moved and connection string parsing was changed (https://github.com/golang-migrate/migrate/pull/517#discussion_r591790853), but that's all. I just meant to explain why there were additional code changes accompanying the upgrade from v3 -> v4.

**Review by dhui — APPROVED** [review]
Thanks for the PR and addressing all of the feedback!
If you want pgx bundled in the CLI releases, you'll need to update `DATABASE` in the `Makefile`.

**Comment by dzbee** [comment]
Thanks for the tip, I definitely missed that. And thanks for working with me to get this added! It's a big help.

File: README.md
Comment (dhui) [inline README.md:1]: Please remove the generated TOC. If we add this, we're likely to use a different generator.

**Review by dhui — CHANGES_REQUESTED** [review]

**Review by dhui — APPROVED** [review]

**Pull request merged**
